File: src/mrkto/resources/static_list.py
"""Static list lookups and membership changes."""

from __future__ import annotations

from mrkto.client import MarketoAPIError


def _lead_inputs(lead_ids: list[int]) -> list[dict[str, int]]:
    return [{"id": lead_id} for lead_id in lead_ids]


def _validate_lead_ids(lead_ids: list[int]) -> None:
    if not lead_ids:
        raise ValueError("At least one lead id is required")
    if len(lead_ids) > 300:
        raise MarketoAPIError("invalid_input", "A maximum of 300 leads is allowed per static list request")


def list_static_lists(
    client,
    *,
    name: str | None = None,
    program_name: str | None = None,
    workspace_name: str | None = None,
    limit: int | None = None,
) -> dict:
    params = {}
    if name:
        params["name"] = name
    if program_name:
        params["programName"] = program_name
    if workspace_name:
        params["workspaceName"] = workspace_name
    return client.get_all_pages("/v1/lists.json", params=params or None, limit=limit)


def get_static_list(client, *, list_id: int) -> dict:
    return client.get(f"/v1/lists/{list_id}.json")


def get_static_list_members(
    client,
    *,
    list_id: int,
    fields: str | None = None,
    limit: int | None = None,
) -> dict:
    params = {}
    if fields:
        params["fields"] = fields
    return client.get_all_pages(f"/v1/lists/{list_id}/leads.json", params=params or None, limit=limit)


def add_to_static_list(
    client,
    *,
    list_id: int,
    lead_ids: list[int],
    dry_run: bool = True,
) -> dict:
    _validate_lead_ids(lead_ids)
    body = {"input": _lead_inputs(lead_ids)}
    if dry_run:
        return {
            "dry_run": True,
            "resource": "static-list",
            "action": "add",
            "list_id": list_id,
            "request": body,
        }
    return client.post(f"/v1/lists/{list_id}/leads.json", json_body=body)


def remove_from_static_list(
    client,
    *,
    list_id: int,
    lead_ids: list[int],
    dry_run: bool = True,
) -> dict:
    _validate_lead_ids(lead_ids)
    body = {"input": _lead_inputs(lead_ids)}
    params = {"id": lead_ids}
    if dry_run:
        return {
            "dry_run": True,
            "resource": "static-list",
            "action": "remove",
            "list_id": list_id,
            "request": body,
            "params": params,
        }
    return client.delete(f"/v1/lists/{list_id}/leads.json", params=params, json_body=body)


def check_static_list_membership(client, *, list_id: int, lead_ids: list[int]) -> dict:
    _validate_lead_ids(lead_ids)
    return client.get(f"/v1/lists/{list_id}/leads/ismember.json", params={"id": lead_ids})

File: internal/cmd/static_list.go
package cmd

import (
	"fmt"
	"strconv"

	"github.com/spf13/cobra"
)

func newStaticListCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	staticListCmd := &cobra.Command{
		Use:   "static-list",
		Short: "Static list lookups.",
		RunE: func(cmd *cobra.Command, args []string) error {
			return cmd.Help()
		},
	}

	staticListCmd.AddCommand(newStaticListListCmd(runtime, options))
	staticListCmd.AddCommand(newStaticListGetCmd(runtime, options))
	staticListCmd.AddCommand(newStaticListMembersCmd(runtime, options))
	staticListCmd.AddCommand(newStaticListCheckCmd(runtime, options))

	return staticListCmd
}

func newStaticListListCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	var (
		name          string
		programName   string
		workspaceName string
		limit         int
	)

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List static lists.",
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			params := map[string]any{}
			if name != "" {
				params["name"] = name
			}
			if programName != "" {
				params["programName"] = programName
			}
			if workspaceName != "" {
				params["workspaceName"] = workspaceName
			}

			result, err := apiClient.GetAllPages("/v1/lists.json", paramsOrNil(params), limit, 0)
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}

	flags := cmd.Flags()
	flags.StringVar(&name, "name", "", "Filter by exact static list name.")
	flags.StringVar(&programName, "program", "", "Filter by program name.")
	flags.StringVar(&workspaceName, "workspace", "", "Filter by workspace name.")
	flags.IntVar(&limit, "limit", 0, "Maximum number of records to return.")

	return cmd
}

func newStaticListGetCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	return &cobra.Command{
		Use:   "get <list-id>",
		Short: "Get a static list by id.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			result, err := apiClient.Get("/v1/lists/"+args[0]+".json", nil)
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}
}

func newStaticListMembersCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	var (
		fields string
		limit  int
	)

	cmd := &cobra.Command{
		Use:   "members <list-id>",
		Short: "List static list members.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			params := map[string]any{}
			if fields != "" {
				params["fields"] = fields
			}

			result, err := apiClient.GetAllPages("/v1/lists/"+args[0]+"/leads.json", paramsOrNil(params), limit, 0)
			if err != nil {
				return err
			}

			return writeResultFields(runtime, options, result, parseFields(fields))
		},
	}

	flags := cmd.Flags()
	flags.StringVar(&fields, "fields", "", "Comma-separated fields to return or display.")
	flags.IntVar(&limit, "limit", 0, "Maximum number of records to return.")

	return cmd
}

func newStaticListCheckCmd(runtime *Runtime, options *RootOptions) *cobra.Command {
	var leadIDs []int

	cmd := &cobra.Command{
		Use:   "check <list-id>",
		Short: "Check whether leads belong to a static list.",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(leadIDs) == 0 {
				return fmt.Errorf("At least one lead id is required")
			}
			if len(leadIDs) > 300 {
				return fmt.Errorf("A maximum of 300 leads is allowed per static list request")
			}

			apiClient, err := loadClient(runtime, options.Profile)
			if err != nil {
				return err
			}

			result, err := apiClient.Get("/v1/lists/"+args[0]+"/leads/ismember.json", map[string]any{"id": intsToStrings(leadIDs)})
			if err != nil {
				return err
			}

			return writeResult(runtime, options, result)
		},
	}

	cmd.Flags().IntSliceVar(&leadIDs, "lead", nil, "Lead id to check. Repeat the option for multiple leads.")
	return cmd
}

func paramsOrNil(values map[string]any) map[string]any {
	if len(values) == 0 {
		return nil
	}
	return values
}

func intsToStrings(values []int) []string {
	result := make([]string, 0, len(values))
	for _, value := range values {
		result = append(result, strconv.Itoa(value))
	}
	return result
}
